### ISSUE TITLE
Can I cast types inside Lua?

### ISSUE THREAD

**Issue by snaulX** [kunitoki/LuaBridge3#160]
I'm writing Lua bindings for making _Components_ for some engine. These components are placed inside _Node_. Nodes can be different: audio, triggers, meshes and etc. Inside the engine it's implemented as inheritation: `class Trigger: public Node`.  

I pass to Lua _component_ function `getNode()` that returns `Node` for current _component_. But when I trying to call `Trigger` methods (because this _Node_ is _Trigger_), of course, shows error about "method doesn't exists in this variable". So I'm thinking about doing in lua something like `node:as(Trigger):methodOfTrigger()`. But I can't imagine how.

How can I cast types inside Lua? Need I to make function for every possible type casting like `node:asTrigger()` where `asTrigger` will implemented in C++ like `LuaRef node; node.cast<Trigger>().value();`?

TL;DR I pass Node and Trigger to lua using `beginClass<Node>("Node")` and `deriveClass<Trigger, Node>("Trigger")`

**Comment by kunitoki** [comment]
It is not possible for LuaBridge to perform automatic downcasting, for the same reason as this is not possible in C++ without an explicit dynamic_cast. You will have to add an explicit cast method:

```cpp

namespace {
struct NodeX
{
    virtual ~NodeX() = default;
};

struct TriggerX : NodeX
{
    int trigger() const
    {
        return 2;
    }
};
} // namespace

TEST_F(ClassTests, DerivedClassCast)
{
    TriggerX objectInstance;

    luabridge::getGlobalNamespace(L)
        .beginClass<NodeX>("NodeX")
        .endClass()
        .deriveClass<TriggerX, NodeX>("DerivedX")
            .addFunction("trigger", &TriggerX::trigger)
        .endClass()
        .addFunction("getNode", [&]() -> NodeX* { return std::addressof(objectInstance); })
        .addFunction("asTrigger", [](NodeX* obj) { return dynamic_cast<TriggerX*>(obj); });

    ASSERT_TRUE(runLua("node = getNode (); result = asTrigger(node):trigger()"));
    ASSERT_TRUE(result().isNumber());
    EXPECT_EQ(result().cast<int>(), 2);
}
```

**Comment by kunitoki** [comment]
You can also register a table for casting using registered types instead of a custom method or using types as strings:

```cpp

namespace {
struct NodeX
{
    virtual ~NodeX() = default;
};

struct TriggerX : NodeX
{
    int trigger() const
    {
        return 2;
    }
};

struct ExplosionX : NodeX
{
    int explode() const
    {
        return 3;
    }
};

template <class F>
void registerCast(lua_State* L, const char* typeName, F&& typeCast)
{
    luabridge::getGlobal(L, typeName).push();

    lua_pushlightuserdata(L, reinterpret_cast<void*>(typeCast));
    luabridge::lua_pushcclosure_x(L, +[](lua_State* L) -> int
    {
        auto typeCast = reinterpret_cast<F>(lua_touserdata(L, lua_upvalueindex(1)));

        luabridge::LuaRef castedObject(L, typeCast(luabridge::LuaRef::fromStack(L, 1)));
        castedObject.push();

        return 1;
    }, 1);

    lua_settable(L, -3);
};
} // namespace

TEST_F(ClassTests, DerivedClassCast)
{
    TriggerX triggerInstance;
    ExplosionX explosionInstance;

    luabridge::getGlobalNamespace(L)
        .beginClass<NodeX>("NodeX")
        .endClass()
        .deriveClass<TriggerX, NodeX>("TriggerX")
            .addFunction("trigger", &TriggerX::trigger)
        .endClass()
        .deriveClass<ExplosionX, NodeX>("ExplosionX")
            .addFunction("explode", &ExplosionX::explode)
        .endClass()
        .addFunction("getNode", [&](int value) -> NodeX*
        {
            if (value == 0)
                return std::addressof(triggerInstance);
            
            return std::addressof(explosionInstance);
        });

    luabridge::LuaRef castTable = luabridge::newTable(L);
    castTable.push();
    registerCast(L, "TriggerX", +[](luabridge::LuaRef base) { return dynamic_cast<TriggerX*>(base.cast<NodeX*>().value()); });
    registerCast(L, "ExplosionX", +[](luabridge::LuaRef base) { return dynamic_cast<ExplosionX*>(base.cast<NodeX*>().value()); });
    castTable.pop();
    luabridge::setGlobal(L, castTable, "cast");

    ASSERT_TRUE(runLua("node = getNode (0); result = cast[TriggerX](node):trigger()"));
    ASSERT_TRUE(result().isNumber());
    EXPECT_EQ(result().cast<int>(), 2);

    ASSERT_TRUE(runLua("node = getNode (1); result = cast[ExplosionX](node):explode()"));
    ASSERT_TRUE(result().isNumber());
    EXPECT_EQ(result().cast<int>(), 3);
}
```

**Comment by snaulX** [comment]
Thanks for the answer. But I was developed own solution too (I forget about this issue to post it here). I wrote two functions:
```cpp
/// Implementing `object:as(Type)`
static int l_as(lua_State* L) {
    luabridge::LuaRef from = luabridge::LuaRef::fromStack(L, 1);
    luabridge::LuaRef type = luabridge::LuaRef::fromStack(L, 2);
    luabridge::LuaResult result = type["cast"](from);
    if (!result) {
        luabridge::raise_lua_error(L, "Calling of 'cast' in 'as' got error: %s", result.errorMessage().c_str());
        return 0;
    }
    luabridge::LuaRef returned = result[0];
    returned.push();
    return 1;
}

/// Implementing helper function `To_Type.cast(object_from)` for `l_as`
template<typename To, typename From>
static int l_cast(lua_State* L) {
    luabridge::TypeResult<From*> from = luabridge::Stack<From*>::get(L, 1);
    if (!from) {
        luabridge::raise_lua_error(L, "'from' value doesn't match type in 'cast' function: %s", from.message().c_str());
        return 0;
    }
    To* to = static_cast<To*>(from.value());
    luabridge::Result result = luabridge::Stack<To*>::push(L, to);
    if (!result) {
        luabridge::raise_lua_error(L, "Pushing casted value in 'cast' got error: %i", result.error());
        return 0;
    }
    return 1;
}
```
and add this to bindings declaration
```cpp
.beginClass<Node>("Node")
    .addFunction("as", l_as)
.endClass()
.deriveClass<Trigger, Node>("Trigger")
    .addStaticFunction("cast", l_cast<Trigger, Node>)
.endClass()
```
and in result Lua code looks like
```lua
node:as(Trigger):trigger()
-- which is equivalent to
Trigger.cast(node):trigger()
```

**Comment by kunitoki** [comment]
Static casting like that is risky. If your node pointer is not a `To` you will end up with undefined behaviour (imagine your node is not a Trigger)
